pkg/datagatherer/versionchecker/versionchecker.go

```diff
@@ -0,0 +1,255 @@
+package versionchecker
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/jetstack/preflight/pkg/datagatherer"
+	"github.com/jetstack/preflight/pkg/datagatherer/k8s"
+	vcapi "github.com/jetstack/version-checker/pkg/api"
+	vcclient "github.com/jetstack/version-checker/pkg/client"
+	selfhosted "github.com/jetstack/version-checker/pkg/client/selfhosted"
+	vcchecker "github.com/jetstack/version-checker/pkg/controller/checker"
+	vcsearch "github.com/jetstack/version-checker/pkg/controller/search"
+	vcversion "github.com/jetstack/version-checker/pkg/version"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// Config is the configuration for a VersionChecker DataGatherer.
+type Config struct {
+	// the version checker dg will also gather pods and so has the same options
+	// as the dynamic datagatherer
+	Dynamic                      k8s.ConfigDynamic
+	VersionCheckerClientOptions  vcclient.Options
+	VersionCheckerCheckerOptions vcapi.Options
+}
+
+// UnmarshalYAML unmarshals the ConfigDynamic resolving GroupVersionResource.
+func (c *Config) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	aux := struct {
+		Dynamic struct {
+			KubeConfigPath    string   `yaml:"kubeconfig"`
+			ExcludeNamespaces []string `yaml:"exclude-namespaces"`
+			IncludeNamespaces []string `yaml:"include-namespaces"`
+		} `yaml:"k8s"`
+		Registries []struct {
+			Kind   string            `yaml:"kind"`
+			Params map[string]string `yaml:"params"`
+		} `yaml:"registries"`
+	}{}
+	err := unmarshal(&aux)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal version checker config: %s", err)
+	}
+
+	c.Dynamic.KubeConfigPath = aux.Dynamic.KubeConfigPath
+	c.Dynamic.ExcludeNamespaces = aux.Dynamic.ExcludeNamespaces
+	c.Dynamic.IncludeNamespaces = aux.Dynamic.IncludeNamespaces
+	// gvr must be pods for the version checker dg
+	c.Dynamic.GroupVersionResource.Group = ""
+	c.Dynamic.GroupVersionResource.Version = "v1"
+	c.Dynamic.GroupVersionResource.Resource = "pods"
+
+	c.VersionCheckerClientOptions.Selfhosted = map[string]*selfhosted.Options{}
+	registryKindCounts := map[string]int{}
+	for i, v := range aux.Registries {
```

Comment:
I'm keen to leave it open to having multiple registries of the same kind in the future (as is possible with the self hosted configs).

I'd have done as you said if it were easier to update the agent configs, but when it's installed in many customer clusters I felt it was better to go with something that worked in both cases.

Reply:
I'm a bit dubious of this being a real use case, and would have thought a single sufficiently scoped identity for each cluster would be enough. I haven't come across a customer whose needed separate service accounts for the same registry provider, or indeed wasn't in a position to create a new one which could be used just for version-checker.

>  but when it's installed in many customer clusters I felt it was better to go with something that worked in both cases.

Would this not be a case of having different value files for different clusters, which is fairly standard anyway.

I don't plan on adding this functionality to version-checker unless there was a real call for it in an issue by multiple people. It is also a nightmare to get it to work programmatically- you would at worst, try each client until one worked, and at best make the user need to define exactly which images are for which client, even though they share the same host.

Reply:
I added a comment about this in 455d035980a282cdd824b5d49152faece38eef9e

Reply:
I'm mostly doing this as we've been burned by it before in preflight. We used to only allow one datagather of each kind for example. Since then, I've generally preferred to have a list rather than a more rigid map for to make it more flexible with little overhead. I think the overhead is worth it to me when the alternative is updating agent config which slows us down a lot as we need to get users to update and have no means of automating it yet.

w.r.t not adding it to VC. Fair enough. My only comment on would be related to this:

> Would this not be a case of having different value files for different clusters, which is fairly standard anyway.

At the moment our model is one agent per cluster and that'd the route we guide users down too. However, everything else in preflight has been designed and built to support multiple clusters and environments if the user wanted to do that. At one point there was talk about running the agent in cloud run for example (talking to many clusters/envs).

Perhaps that's something we need to revisit, but it's an assumption I continue to make in preflight for the time being.